### DISCUSSION TITLE
How to render tbl_summary (or any other table) to word and being able to reference it

### DISCUSSION THREAD

**Discussion by siganna71** [quarto-dev/quarto-cli#6672] [Q&A]
### Description

I am using the tbl_summary function and I want to be to reference my table when I render to Word. The reference works fine when I render to PDF. I am using the function as_flex_table() to print the table but I am willing to use some other package if that helps. So how can I add caption to a table and reference it when I render to Word?

With thanks
Anna

**Comment by mcanouil** [comment]
Could you share a small self-contained "working" (reproducible) example to work with, _i.e._, a complete Quarto document or a Git repository? Thanks.

You can share a Quarto document using the following syntax, _i.e._, using more backticks than you have in your document (usually four ` ```` `).

`````md
````qmd
---
title: "Reproducible Quarto Document"
format: html
---

This is a reproducible Quarto document using `format: html`.
It is written in Markdown and contains embedded R code.
When you run the code, it will produce a plot.

```{r}
plot(cars)
```

The end.
````
`````

**Reply by siganna71** [reply]
I will do that. Why should I have more backticks?

**Reply by mcanouil** [reply]
To escape the ones inside.
``` `` ` `` ``` produces `` ` ``

**Comment by siganna71** [comment]
Something like this ok?
````qmd

---
title: "Testing referencing tables"
format: docx
---
 I want to be able to reference  @tbl-4a. It works fine if I choose format "pdf" but not for doxs
```{r, echo = FALSE, warning = FALSE, message = FALSE}
library(tidyverse)
library(ggplot2)
library(gtsummary)
library(flextable)

my_iris <- iris
```

```{r, echo = FALSE, warning = FALSE, message = FALSE}
tbl_species <- tbl_summary(
 my_iris,
  by = Species,
  type = list(c(Sepal.Length, Sepal.Width, Petal.Length, Petal.Width) ~ 'continuous'),  # Makes it treat LEC as continuous variable
  statistic = list(all_continuous() ~ "{mean} ({sd})"),
  digits = all_continuous() ~ 1,
  include = c(Sepal.Length, Sepal.Width, Petal.Length, Petal.Width),
  missing_text = "Missing") |> 
  add_p(test = list(all_continuous() ~ "oneway.test",  
                    all_categorical() ~ "chisq.test")) |> 
  bold_labels() |> 
  modify_header(label = "", p.value = "**P**", all_stat_cols() ~ "**{level}**, \n N = {n}") 
```

```{r, echo = FALSE, warning = FALSE, message = FALSE}
#| label: tbl-4a
#| tbl-cap: "Characteristics of SAGA participants stratified by selfinjury"

tbl_species %>% as_flex_table() 
```

````

**Reply by cderv** [reply]
@rich-iannone do you known if there is any issue with **gt** docx output and reference ? 

Simpler example

````markdown
---
title: "Testing referencing tables"
format: docx
---

I want to be able to reference  @tbl-4a. It works fine if I choose format "pdf" but not for doxs

```{r}
#| label: tbl-4a
#| tbl-cap: "Characteristics of SAGA participants stratified by selfinjury"
gt::gt(mtcars)
```
````

Possibly related to 
* https://github.com/rstudio/gt/issues/1140

@cscheid maybe another test for your work on Cross ref ?

**Reply by mcanouil** [reply]
Yes, it feels the issue comes from `gtsummary` that outputs something weird with its `as_flextable` (and `as_gt` functions so maybe also an issue in `gt`).

**Reply by cderv** [reply]
Thanks I missed the `as_flex_table()` 🤷‍♂️  - got stuck with the name **gt**summary 😅 

So definitely that as cross ref and flextable is not yet working

* https://github.com/quarto-dev/quarto-cli/issues/1556
* https://github.com/davidgohel/flextable/issues/494

**Reply by siganna71** [reply]
Thank you all for taking the time to answer. I am willing to use some other package than flextable. But based on what I have been reading in those links you provided, there is no solution at the moment, right? Do you think this will be fixed in near future?

**Reply by cderv** [reply]
What is the desired output exactly ? 

You could output your gtsummary to a `kable` maybe and this would be Markdown table that would be rendered in Docx too. 

Or maybe render to HTML table, that Quarto would also parse like a Markdown table, and output the correct one for Docx. 
https://quarto.org/docs/authoring/tables.html#html-tables

I believe for now, package that output raw openxml to be included in docx output are not working correctly with Quarto. To me it should be solved soon - and it may even just be a tweak in the package function itself.

**Reply by siganna71** [reply]
The desired output is to be able to reference tables in my text when I render to word. I have had some trouble with kable and Word but maybe I should try that again. 
I have never tried  to render to HTML , I will look into that.
Hope you will fix the interaction between raw openxml and Quarto

**Reply by cderv** [reply]
Did you try `as_kable()` in **gtsummary** ? 

````
```{r, echo = FALSE, warning = FALSE, message = FALSE}
#| label: tbl-4a
#| tbl-cap: "Characteristics of SAGA participants stratified by selfinjury"

tbl_species %>% as_kable() 
```
````

**Reply by siganna71** [reply]
I tried as_kable() . The reference works fine but I lose all the nice formatting from tbl_summary

**Reply by cderv** [reply]
Ah sure - for the formatting in Docx, it needs support for flextable or gt with cross reference. 

Follow the linked issue already mentioned

**Reply by e-linc** [reply]
Could you elaborate on this, please? I'm new to Quarto and gtsummary, and encountered the same crossref problem when outputting my `tbl_summary()` table to .docx. As @siganna71 reported, when converting the table via `as_kable()`, the crossref worked, but the table formatting was lost in the .docx file. Is there an updated solution for this now?

**Reply by mcanouil** [reply]
@e-clin could you share a small self-contained "working" (reproducible) example to work with, _i.e._, a complete Quarto document or a Git repository? Thanks.

You can share a Quarto document using the following syntax, _i.e._, using more backticks than you have in your document (usually four ` ```` `).

`````md
````qmd
---
title: "Reproducible Quarto Document"
format: html
engine: knitr
---

This is a reproducible Quarto document using `format: html`.
It is written in Markdown and contains embedded R code.
When you run the code, it will produce a plot.

```{r}
plot(cars)
```

![A placeholder image](https://placehold.co/600x400.png)

The end.
````
`````

**Reply by e-linc** [reply]
Thanks very much for your quick reply. Please see my sample code below:

````qmd
---
title: "Testing gtsummary table crossref in Word output"
format: docx
execute:
  echo: false
  warning: false
---

```{r}
#| label: load-packages
#| include: false

library(tidyverse)
library(gtsummary)
library(flextable)
```

## Results

@tbl-mtcars-summary summarises the cars in the mtcars dataset by transmission.

```{r}
#| label: tbl-mtcars-summary
#| tbl-cap: Summary of cars in mtcars by transmission.

mtcarsSummaryTbl <- mtcars %>% 
  tbl_summary(by = am) %>% 
  modify_header(
    stat_1 = "**Automatic**, N = {n}",
    stat_2 = "**Manual**, N = {n}",
  ) %>% 
  add_overall() %>% 
  bold_labels()

mtcarsSummaryTbl
```

````

The above (I believe is through flextable?) produced the nicest-looking table in the .docx file, but crossref didn't work:

![test-gtsummary-table-word-crossref-as-flextable](https://github.com/quarto-dev/quarto-cli/assets/26250149/4cb010ae-b516-45a1-ba4b-13d83620fdf0)

I tried `as_kable()`, which fixed the crossref, but lost some table formatting:

![test-gtsummary-table-word-crossref-as-kable](https://github.com/quarto-dev/quarto-cli/assets/26250149/65d4a368-a552-4593-a8fd-c1f17d7547d8)

I also tried `as_gt()`, which didn't generate a pretty table, nor did the crossref work:

![test-gtsummary-table-word-crossref-as-gt](https://github.com/quarto-dev/quarto-cli/assets/26250149/9d3441d5-e697-4973-bab0-c59177eb55cb)


.

**Reply by cderv** [reply]
@e-clin thanks for sharing. 

> The above (I believe is through flextable?)

Yes flextable will be the default for docx document. See their doc: https://www.danieldsjoberg.com/gtsummary/articles/rmarkdown.html

First for anyone who tries with latest version tools, there is currently an issue reported at https://github.com/davidgohel/officer/issues/546

And unfortunately we have also known issue with a docx validation problem : 
- https://github.com/quarto-dev/quarto-cli/issues/7321

and unfortunately **gt** may have the same type of issue:
- https://github.com/quarto-dev/quarto-cli/issues/7151

so some limitations for now that we are trying to fix.

> I tried as_kable(), which fixed the crossref, but lost some table formatting:

With **kable**, it will produce a Markdown tables, and you won't get the same formatting feature as what is possible with **gt** or **flextable**

**Reply by e-linc** [reply]
@cderv Thanks very much for summarising the current situation. I went through the links but none of them seemed to be fully resolved. Am I right in understanding that for now, it's best to use **kable** for the crossref and table caption to work in the Word output, with a little sacrifice in table formatting?

**Reply by cderv** [reply]
yes this seems right. Otherwise you need to use **gt** or **flextable** to create a nice table to save as an image. I believe they have some  export function. You'll get more formatting in your table, but it won't be a table in DOCX anymore. 

Overall, not great situation, we'll fix it ! openxml is a bit harder to debug than other stuff - so it takes a bit longer sorry.

**Reply by e-linc** [reply]
Thanks very much for your suggestions. No problem - I can manage with **kable** for now. Thanks for the great work on Quarto! I just started using it to write a paper for the first time, so am still learning a lot of things. Your fix may come sooner than I figure out how to revise in Quarto the Word documents with tracked changes from co-authors. ;)

Looking forward to the update.
